site/docs/snapshot-tag-branch.md

```diff
@@ -0,0 +1,148 @@
+
+#### Snapshot reference retention policy
+
+Similarly, snapshot reference has the properties below to provider finer grain control:
+
+| Property                     | Type      | Description |
+|------------------------------|-----------|-------------|
+| **`min-snapshots-to-keep`**  | `int`     | For `branch` type only, the minimum number of snapshots to keep in a branch |
+| **`max-snapshot-age-ms`**    | `long`    | The duration before a snapshot tagged or in a branch could be expired by any automatic snapshot expiration process  |
+
+#### Policy evaluation mechanism
+
+When a snapshot expiration process starts, it follows the steps described below:
+
+1. form an expiration candidate pool containing all snapshots
+2. for each snapshot reference, evaluate the associated policy and move snapshots out of the candidate pool
+3. apply global retention policy to and move snapshots out of the candidate pool
+4. when multiple snapshots can be chosen to be moved out, newer snapshots win
+4. after evaluation, expire all snapshots that are still in the candidate pool
+
+#### Policy evaluation example
+
+Here is an example for how an Iceberg snapshot expiration procedure evaluates what snapshots to expire.
+
+Suppose we have the following snapshot graph and retention policies configured:
+
+```
+A -> B -> C (main)
+ \      (dev)
+  D -> E (b1)
+   \
+    F -> G (b2)
+```
+
+| Policy Type      | Max Age       | Min to Keep | Snapshots Affected    |
+|------------------|---------------|-------------|-----------------------|
+| global           | 5 hours       | 4           | A, B, C, D, E, F, G   |
+| branch/main      | 3 days        | 1           | A, B, C               |
+| branch/b1        | 2 days        | 2           | A, D,  E              |
+| branch/b2        | 1 day         | 0           | A, D, F, G            |
+| tag/dev          | forever       | N/A         | C                     |
+
+Assume that we have a process continuously running the snapshot expiration procedure, we would have the results below as time progresses:
+
+##### Day 1: F and G are expired
+
+On day 1, the global and branch b2 max age has passed, affecting A, D, F, G.  
+A andD cannot be expired due to branch b1 policy, so only F and G are expired.
+
+``` 
+A -> B -> C (main)
+ \      (dev)
+  D -> E (b1)
+```
+
+##### Day 2: no snapshot is expired
+
+On day 2, branch b1 max age has also passed, affecting A, D, E.
+Because branch b1 must keep 2 snapshots, only the oldest snapshot of the branch A could be expired. 
+However, A is also in the main branch, so it cannot be expired either.
+As a result, no snapshot is expired.
+
+``` 
+A -> B -> C (main)
+ \      (dev)
+  D -> E (b1)
+```
+
+##### Day 3: A is expired
+
+On day 3, main branch max age has also passed, affecting A, B, C.
+Based on the global policy, we must keep 4 snapshots, so can only expire 1 snapshot, 
```

Comment:
Do we really need to honour the global policy when that branch is having local policy ? Is it better to delete A and B in this case as local policy says min snapshots to keep is 1 ?
I think applying global policies is good enough only for snapshots that are not involved in any branch, tag. WDYT ?

Reply:
Yes, based on the mechanism described to determine snapshots to expire, we are determining which snapshots are not removed after evaluating all the related policies, so it would eventually hit the global policy and stop there.

I think it also fits the meaning of "minimum number of snapshots to keep", because it's the minimum, not that we need to always reach that value.

Reply:
Another reason to avoid saying "global policy" is that it isn't clear whether branches override "global" or whether there is some way to merge them (like max(table-min-snapshots, branch-min-snapshots)).

I think calling the "global" settings "table defaults" is a good way to help people understand.